sources/group.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Odyssey.
+ *
+ * Scalable PostgreSQL connection pooler.
+ */
+
+#include <kiwi.h>
+#include <machinarium.h>
+#include <odyssey.h>
+
+int od_group_free(od_group_t *group)
+{
+	if (group == NULL)
+		return NOT_OK_RESPONSE;
+
+	if (group->route_usr)
+		free(group->route_usr);
+	if (group->route_db)
+		free(group->route_db);
+	if (group->storage_user)
+		free(group->storage_user);
+	if (group->storage_db)
+		free(group->storage_db);
+	if (group->group_name)
+		free(group->group_name);
+	if (group->group_query)
+		free(group->group_query);
+
+	free(group);
+	return OK_RESPONSE;
+}
+
+void od_group_qry_format(char *qry, char *fmt, ...)
+{
+	va_list args;
+	va_start(args, fmt);
+	int len = od_vsnprintf(qry, OD_QRY_MAX_SZ, fmt, args);
+	va_end(args);
+
+	/* dirty hack */
+	qry[len] = '\0';
+}
+
+int od_group_parse_val_datarow(machine_msg_t *msg, int *is_group_member)
+{
+	char *pos = (char *)machine_msg_data(msg) + 1;
+	uint32_t pos_size = machine_msg_size(msg) - 1;
+
+	/* size */
+	uint32_t size;
+	int rc;
+	rc = kiwi_read32(&size, &pos, &pos_size);
+	if (kiwi_unlikely(rc == -1))
+		goto error;
+	/* count */
+	uint16_t count;
+	rc = kiwi_read16(&count, &pos, &pos_size);
+
+	if (kiwi_unlikely(rc == -1))
```

Comment:
may be kiwi_likely here?

or if "kiwi_likely" does not exist write rc != -1

Reply:
why is this better?

Reply:
because it will works))
I mean write kiwi_likely(rc == -1) or kiwi_unlikely(rc != -1)

because in your case if rc is equal to -1 (error happened) => rc == -1 - true (1) => kiwi_unlikely returns false (because :#define kiwi_unlikely(expr) __builtin_expect(!!(expr), 0)), so you won't go into if block (so we won't do goto error)

Reply:
everything is correct in the code. if expr "rc == -1" is false, than kiwi_unlikely return false else return true

Reply:
if expo "rc == -1" is false --> !!(rc == -1) == 0 --> kiwi_unlikeky returns 

#define kiwi_unlikely(expr) __builtin_expect(!!(expr), 0) -> 

#define kiwi_unlikely(expr) __builtin_expect(0, 0) -> true